### PR TITLE
First commit for the creation of a qkeras zoo. 

### PR DESCRIPTION
**Pull request by FrancescoLoro**
This folder contains a collection of networks written using two different frameworks: qkeras and larq. 
Each network can be built and tested using a randomly generated dataset, the output will consist of two measurements:
- Mean MSE, calculate the mean MSE between all the output for both networks
- Absolute errors, calculates how many times the class predicted by one network does
    not coincide with the class predicted by the other network

The folder is divided in:
- network_name.py is the class to build and test the networks.
- utils.py contains share methods between all the classes.
- results.txt are the results of the comparison between the networks.
- ./qkeras_models contains the .json qkeras saved models.
- ./larq_models contains the .json larq saved models.
- ./weights please put the downloaded weights here, link to download weights is provided in the network class.

Link to the folder with all weights: https://drive.google.com/drive/folders/1pGZ6dGWvJyc9aH-TOQohm0PhORihQZ5I?usp=sharing

ALready added networks: quicknet, quicknet_small, quicknet_large, alexnet, and birealnet.

### CODE REVIEW

File: zoo/alexnet.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+import larq as lq
+from utils import compare_network, create_random_dataset, dump_network_to_json
+
+# Define path to the pre-trained weights
+path_alexnet = "./weigths/binary_alexnet_weights.h5"
+alexnet_name = "alexNet"
+
+
+class alexnet():
+  """
```
Comment (zhuangh) [inline zoo/alexnet.py:33]: could you follow https://google.github.io/styleguide/pyguide.html#384-classes, also apply for other places as well, thanks!

File: zoo/alexnet.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+import tensorflow as tf
+import larq as lq
+from utils import compare_network, create_random_dataset, dump_network_to_json
+
+# Define path to the pre-trained weights
+path_alexnet = "./weigths/binary_alexnet_weights.h5"
+alexnet_name = "alexNet"
+
+
+class alexnet():
```
Comment (zhuangh) [inline zoo/alexnet.py:32]: remove "()"
and also follow https://google.github.io/styleguide/pyguide.html#3162-naming-conventions

File: zoo/alexnet.py (hunk 1 above)
Comment (FrancescoLoro) [reply zoo/alexnet.py:33]: Yes, sure!

File: zoo/alexnet.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+# Define path to the pre-trained weights
+PATH_ALEXNET = "weights/binary_alexnet_weights.h5"
+ALEXNET_NAME = "alexNet"
+
+
+class AlexNet:
+  """
+  Class to create and load weights of: alexnet
+  Attributes:
+        network_name: Name of the network
```
Comment (zhuangh) [inline zoo/alexnet.py:37]: one sentence in the first line of the docstring followed by one empty line

ref: https://google.github.io/styleguide/pyguide.html#384-classes

could you use 2 or 4 spaces for the indentation? thanks

File: zoo/alexnet.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  """
+
+  def __init__(self):
+    self.__weights_path = PATH_ALEXNET
+    self.network_name = ALEXNET_NAME
+
+  @staticmethod
+  def add_qkeras_conv_block(model, filters_num, kernel_size, pool,
+                            qnt, strides=1):
+    """
```
Comment (zhuangh) [inline zoo/alexnet.py:47]: Move one sentence in the first line followed by an empty line.

ref: https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

File: zoo/birealnet.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                        kernel_quantizer="binary(alpha=1)", use_bias=False))
+    model.add(tf.keras.layers.BatchNormalization())
+
+  @staticmethod
+  def add_qkeras_connection_block(model, filters_num):
+    """
+    Add a sequence of: Activation quantization, Quantized Conv2D, reshape,
+    Average Pooling, Conv2D, 2x BatchNormalization
+    :param model: model where to add the sequence
+    :param filters_num: number of filters for Cov2D
```
Comment (zhuangh) [inline zoo/birealnet.py:63]: Cov2D?

File: zoo/birealnet.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+      self.add_qkeras_residual_block(qkeras_biRealNet, 128)
+    self.add_qkeras_connection_block(qkeras_biRealNet, 256)
+
+    for _ in range(0, 3):
+      self.add_qkeras_residual_block(qkeras_biRealNet, 256)
+    self.add_qkeras_connection_block(qkeras_biRealNet, 512)
+
+    for _ in range(0, 3):
+      self.add_qkeras_residual_block(qkeras_biRealNet, 512)
+    qkeras_biRealNet.add(tf.keras.layers.AveragePooling2D(pool_size=(7, 7)))
```
Comment (zhuangh) [inline zoo/birealnet.py:159]: Any reason of not using power of two numbers for pool_size? For example, (2,2) or (4,4)

We need to be careful here regarding the precision, if the pool_size is not power of two (49=7x7 here). 

```avg = sum(pool_element) / pool_size = sum(pool_element) / 49```, which introduces the precision loss if we use fixed point or integer to hold avg, right.

File: zoo/birealnet.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    shape_out = (model.output_shape[1], model.output_shape[2],
+                 model.output_shape[3] // 2)
+    model.add(tf.keras.layers.Flatten())
+    model.add(tf.keras.layers.Reshape(target_shape=(shape_in, 1)))
+    model.add(tf.keras.layers.AvgPool1D(1, strides=2, padding="same"))
+    model.add(tf.keras.layers.Reshape(target_shape=shape_out))
+    model.add(tf.keras.layers.Conv2D(filters_num, (1, 1), padding="same",
+                                     use_bias=False))
+    model.add(tf.keras.layers.BatchNormalization())
+    model.add(tf.keras.layers.BatchNormalization())
```
Comment (zhuangh) [inline zoo/birealnet.py:80]: why consecutive batchnormalization layers?

File: zoo/birealnet.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+import tensorflow as tf
+import larq as lq
+from utils import compare_network, create_random_dataset, dump_network_to_json
+
+# Define path to the pre-trained weights
+PATH_BIREALNET = "weights/birealnet_weights.h5"
+BIREALNET_NAME = "biRealNet"
+
+
+class BirealNet:
```
Comment (zhuangh) [inline zoo/birealnet.py:33]: is it from this paper https://arxiv.org/pdf/1808.00278.pdf

**Review by zhuangh — CHANGES_REQUESTED** [review]
@FrancescoLoro thank you so much for the license modification. 

I added some comments to help us better understand the benchmark models.

FYI, we will discuss the larq part internally first in this week to see whether we want to include the part in this PR. 

BTW, please also try to apply [google python style](https://google.github.io/styleguide/pyguide.html) in the code for readability. thanks!

File: zoo/birealnet.py (hunk 8 above)
Comment (FrancescoLoro) [reply zoo/birealnet.py:33]: Yes, it is

File: zoo/birealnet.py (hunk 5 above)
Comment (FrancescoLoro) [reply zoo/birealnet.py:63]: Apologize, it was a typo

File: zoo/birealnet.py (hunk 7 above)
Comment (FrancescoLoro) [reply zoo/birealnet.py:80]: In next commit I've done a complete refactor of that class, fixing also that mistake

File: zoo/birealnet.py (hunk 8 above)
Comment (FrancescoLoro) [reply zoo/birealnet.py:33]: Yes, it is

File: zoo/alexnet.py (hunk 4 above)
Comment (FrancescoLoro) [reply zoo/alexnet.py:47]: Ok
